### PR TITLE
add log control verbs to busctl

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by boucman** [comment]
I'm not convinced myself, but I might as well ask the question...

Wouldn't that make more sense as a systemd-analyze verb ?
Pros and cons:

* this is supposed to be generic and systemd-analyze is systemd specific
* systemd-analyze already has verbs (like cat-config) that are usefull beyond systemd
* buctl is pure-dbus, but this is specific to a given interface... (or maybe the interface should be added to the well-known interfaces of the dbus specification ?)

hmm...

**Comment by fbuihuu** [comment]
yeah I was just wondering myself why it's been added to busctl rather than systemctl.

@boucman you asked for systemd-analyze instead but IIRC `log-level` verb was added to systemctl because systemd-analyze is more a debug tool and setting log-level for units or PID1 is more like a general operation.

**Comment by DaanDeMeyer** [comment]
Maybe put a `log-level` verb on each of the command line tools instead that sets the log-level for that service (like `systemctl` currently does for PID1)? That streamlines the most commen use cases. For services without a command line tool or those not bundled with systemd, `busctl` can still be used.

Putting the verb on the corresponding tool increases the discoverability and ease-of-use quite a bit imo since you don't have to switch to a different tool and you don't have to figure out the name of the service on the bus (which is easy to remember once you've looked it up but not trivial if you've never looked at D-Bus before and are just debugging a service)

**Comment by poettering** [comment]
> Maybe put a `log-level` verb on each of the command line tools instead that sets the log-level for that service (like `systemctl` currently does for PID1)? That streamlines the most commen use cases. For services without a command line tool or those not bundled with systemd, `busctl` can still be used.

Not a fan of redundancy. It just blows up the user interface if we make everything available everywhere... Redundancy in the UI comes at a price. A simple, lean UI will become large and complex if everything is available through three "easy" ways...

**Comment by poettering** [comment]
> Wouldn't that make more sense as a systemd-analyze verb ?

Everything in systemd-analyze is specific to pid1, i.e. we test stuff associated with PID 1. This is different for this operation, it's generic, applies to any dbus service that wants to implement this... Hence I think busctl is the better place for this.

**Comment by poettering** [comment]
> @boucman you asked for systemd-analyze instead but IIRC `log-level` verb was added to systemctl because systemd-analyze is more a debug tool and setting log-level for units or PID1 is more like a general operation. Therefore users expected this command to be available from systemctl too.

I still believe the log level should not be visible in systemctl. It's a debugging tool, nothing else.

**Comment by poettering** [comment]
> buctl is pure-dbus, but this is specific to a given interface... (or maybe the interface should be added to the well-known interfaces of the dbus specification ?)

yeah, i think it should be conceptually ultimately be more part of something like the dbus spec than of our stuff. it isn't right now. but i think it's where it ultimately belongs...

**Comment by boucman** [comment]
maybe discuss this with the dbus people then, so we can have their feedback (I have no contact with that particular project, but that would make sense...)

**Comment by fbuihuu** [comment]
> I still believe the log level should not be visible in systemctl. It's a debugging tool, nothing else.

But since it's been added to systemctl since commit 38fcb7f766c84736425e86854b8a4468c126dafa, and I agree with its reasoning, that would be sad to ignore systemctl.

File: man/busctl.xml [hunk 1]
```diff
@@ -149,6 +149,47 @@
         parameters formatted as strings.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><command>log-level</command> <arg choice="plain"><replaceable>SERVICE</replaceable></arg> <arg choice="opt"><replaceable>LEVEL</replaceable></arg></term>
+
+        <listitem><para>Retrieve or set the maximum log level (a.k.a maximum log priority) of the specified
+        service. This function permits log level control of D-Bus services that implement the generic
```
Comment (keszybz) [inline man/busctl.xml:156]: "permits" seems misplaced. Maybe "allows".

File: man/busctl.xml [hunk 2]
```diff
@@ -149,6 +149,47 @@
...
+        <literal>org.freedesktop.LogControl1</literal> interface. If no level argument is specified retrieves
```
Comment (keszybz) [inline man/busctl.xml:157]: This is (or will be soon) documented in org.freedesktop.LogControl1(5), so please add a ref.

"If no level argument is specified, retrieves"

File: man/busctl.xml [hunk 3]
```diff
@@ -149,6 +149,47 @@
...
+        the current log level, otherwise sets it. The log level should be a typical syslog log level, i.e. a
+        value in the range 0…7 or one of the strings <constant>emerg</constant>, <constant>alert</constant>,
+        <constant>crit</constant>, <constant>err</constant>, <constant>warning</constant>,
+        <constant>notice</constant>, <constant>info</constant>, <constant>debug</constant>; see <citerefentry
+        project='man-pages'><refentrytitle>syslog</refentrytitle><manvolnum>3</manvolnum></citerefentry> for
+        details.</para>
+
+        <para>Note that this requires explicit support in a D-Bus service to work correctly, and typically
+        controls the threshold beyond which all log messages are suppressed. Most service use a maximum log
```
Comment (keszybz) [inline man/busctl.xml:166]: "Most services"

File: man/busctl.xml [hunk 4]
```diff
@@ -149,6 +149,47 @@
...
+        greater importance are written to its log stream, while those of debug priority are
+        suppressed.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><command>log-target</command> <arg choice="plain"><replaceable>SERVICE</replaceable></arg> <arg choice="opt"><replaceable>TARGET</replaceable></arg></term>
+
+        <listitem><para>Retrieve or set the log target of the specified service. This function permits log
+        target control of D-Bus services that implement the generic
+        <literal>org.freedesktop.LogControl1</literal> interface. If no level argument is specified retrieves
```
Comment (keszybz) [inline man/busctl.xml:177]: Comma before "retrieves".

File: man/busctl.xml [hunk 5]
```diff
@@ -149,6 +149,47 @@
...
+        the current log target, otherwise sets it. The log target should be one of the strings
+        <constant>console</constant> (for log output to the service's standard error stream),
+        <constant>kmsg</constant> (for log output to the kernel log buffer), <constant>journal</constant>
+        (for log output to
+        <citerefentry><refentrytitle>systemd-journald.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+        using the native journal protocol), <constant>syslog</constant> (for log output to the classic syslog
+        socket <filename>/dev/log</filename>), <constant>null</constant> (for no log output whatsoever) or
+        <constant>auto</constant> (for an automatically determined choice, typically equivalent to
+        <constant>console</constant> if the service is invoked interactively, and
+        <constant>journal</constant> or <constant>syslog</constant> otherwise).</para>
```
Comment (keszybz) [inline man/busctl.xml:187]: I made this comment in some of the other PRs, but it was never resolved. While the log-level part above is nice and general, the log-target part is very system-manager specific. For most services it simply doesn't make any sense to set anything except "console" or "journal". I'm not sure how to best handle this, but I think it should be somehow emphasized that for most services only some values make sense, and that not implementing any of them is also OK, and e.g. only supporting logging to stdout. I don't want people to get the idea that it is somehow recommended to implement all the options that systemd itself implements for this.

**Comment by poettering** [comment]
This would just be super weird in systemctl, since it operates on a *D-Bus* service name, not a unit name. That really doesn't fit into systemctl. In systemctl the parameters are generally units or unit files or the verbs have no parameter. But mixing in a dbus service name suddenly is just really weird.

**Comment by boucman** [comment]
well, in each xxxctl then... that's probably where it fits the best.

**Comment by poettering** [comment]
@boucman not all our *d have a matching *ctl... and we want this to be implementable by *any* daemon really, not just hours.

**Comment by keszybz** [comment]
> This would just be super weird in systemctl, since it operates on a _D-Bus_ service name, not a unit name.

Right. I this is all that we ever wanted to do with this command, then I'd be against adding it. All it does is saving the user from having to type " /org/freedesktop/LogControl1 org.freedesktop.LogControl1" in the command. Otherwise it's almost equivalent to just `busctl get-property` and 'busctl set-property'.

What we should gun for is an interface where this operates on a *unit* name. And then it belongs in systemctl.

**Comment by keszybz** [comment]
Let's close this one. We have an interface that operates on unit names, and I think it is superior.

**Pull request closed without merging**